### DISCUSSION TITLE
Ensemble retriever with BM25 in realistic settings

### DISCUSSION THREAD

**Discussion by eingrid** [milvus-io/milvus#33849] [Q&A and General discussion]
I have a straightforward query: I'm currently using an Ensemble retriever with a model from HF and BM25SparseEmbedding. In my setup, I'm employing RAG with a non-English language. I find myself fitting BM25SparseEmbedding every time a document gets added to the database, which I'm not particularly fond of because it means recomputing sparse embeddings each time. From what I've gathered, Weaviate doesn't require this. Are there any alternative approaches to avoid this recomputation?

Additionally, I'll include the example code I stumbled upon on Langchain for creating a retriever: https://python.langchain.com/v0.2/docs/integrations/retrievers/milvus_hybrid_search/.

In https://github.com/milvus-io/milvus/discussions/31906, @xiaofan-luan said "you can skip the first step and use pretrained bm25 model as well", maybe someone could tell me how to use pretrained bm25 I guess it will solve my problems

**Comment by yhmo** [comment]
Milvus is a vector database for storing and similar searches for vectors. Vectors should be generated and ready outside the Milvus.
Let's say you have 1 million documents, you can input the 1 million documents to BM25SparseEmbedding to generate 1 million sparse vectors. Then insert the 1 million vectors into Milvus.

**Comment by xiaofan-luan** [comment]
> https://python.langchain.com/v0.2/docs/integrations/retrievers/milvus_hybrid_search/

sparse_embedding_func = BM25SparseEmbedding(corpus=texts)
sparse_embedding_func.embed_query(texts[1])

the example you offered shows how to use bm25 embeddings.

Can you give me more details about  " I find myself fitting BM25SparseEmbedding every time a document gets added to the database"
Did you mean each add also fintune the Bm25 model? I think this is expected behaviour for all BM25 models
